WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,147 @@
+            rootViewController: UIViewController(),
+            alertsPresenter: self,
+            onboardingPresenter: self,
+            bluetoothConnectionController: bluetoothConnectionController,
+            builtInConnectionController: tapToPayConnectionController)
+        let orderPaymentUseCase = await CollectOrderPaymentUseCase(siteID: siteID,
+                                                                   order: order,
+                                                                   formattedAmount: currencyFormatter.formatAmount(order.total, with: order.currency) ?? "",
+                                                                   // moved from EditableOrderViewModel.collectPayment(for: Order)
+                                                                   rootViewController: UIViewController(),
+                                                                   // We don't want to use this at all, but it's currently required by the existing code.
+                                                                   // TODO: replace `rootViewController` with a protocol containing the UIVC functions we need, and implement that here.
+                                                                   onboardingPresenter: self,
+                                                                   configuration: CardPresentConfigurationLoader().configuration,
+                                                                   alertsPresenter: self,
+                                                                   preflightController: preflightController)
+        return await withCheckedContinuation { continuation in
+            orderPaymentUseCase.collectPayment(using: discoveryMethod) { error in
+                if let error = error as? CardPaymentErrorProtocol {
+                    continuation.resume(returning: .failure(error))
+                } else {
+                    continuation.resume(returning: .failure(CardPaymentsAdaptorError.unknownPaymentError(underlyingError: error)))
+                }
+            } onCancel: {
+                continuation.resume(returning: .cancellation)
+            } onPaymentCompletion: {
+                // no-op – not used in PaymentMethodsViewModel anyway so this can be removed
+            } onCompleted: {
+                continuation.resume(returning: .success(order))
+            }
+        }
+    }
+
+    enum CardPaymentsAdaptorError: Error, CardPaymentErrorProtocol {
+        var retryApproach: CardPaymentRetryApproach {
+            .restart
+        }
+
+        case unknownPaymentError(underlyingError: Error)
+    }
+}
+
+extension CardPresentPaymentsAdaptor: CardPresentPaymentsOnboardingPresenting, CardPresentPaymentAlertsPresenting {
+    func showOnboardingIfRequired(from: UIViewController, readyToCollectPayment: @escaping () -> Void) {
+        paymentsAlertHandler?.showOnboarding()
+    }
```

Comment:
> actually try to use it from a view tomorrow. 

Not sure if you've already thought/solved this, but this function would still require to pass a `UIViewController` in, keeping UIKit as dependency. How should we go about it? Some sort of extension to the UIViewController? 

```suggestion
extension UIViewController: Presenter { }

protocol Presenter {
    func present(_ viewControllerToPresent: UIViewController, animated: Bool, completion: (() -> Void)?)
    var navigationController: UINavigationController? { get }
    func show(_ vc: UIViewController, sender: Any?)
}

    func showOnboardingIfRequired(from: Presenter, readyToCollectPayment: @escaping () -> Void) {
        paymentsAlertHandler?.showOnboarding()
    }
```

Problem is that we'll have troubles in SwiftUI when handling this method, perhaps by using generics for the view adapter and a hosting controller? 🤔

Reply:
> this function would still require to pass a UIViewController in, keeping UIKit as dependency. How should we go about it? Some sort of extension to the UIViewController?

The `UIViewController` gets passed in from the existing Card Payments (UIKit-heavy) code that we're adapting... and then we ignore it, so there's no need to pass one in from SwiftUI. 

This is the adaptor impleementing the OnboardingPresenter protocol, for the benefit of the existing payments code, and then translating (adapting) the `showOnboardingIfRequired` call to a more SwiftUI friendly handler. That `paymentsAlertHandler` is what gets passed in from SwiftUI, and it doesn't know about the UIViewController.

Reply:
> still require to pass a `UIViewController` in, keeping UIKit as dependency. How should we go about it? Some sort of extension to the UIViewController?

PS – this is exactly what I intend to do for `rootViewController`, so that I can pass in an adaptor implementation instead of `UIViewController()` as I'm doing at the moment. It may be that we can just ignore whatever calls we get on `rootViewController` when using it via the adaptor, but if we can't, this is the kind of approach I'd take.